### PR TITLE
[Core] Surface In Nurbs Volume Geometry

### PR DESCRIPTION
**Pull request by manuelmessmer**
**Description**
Adds the surface in nurbs volume geometry. This is used to embedded geometries (e.g. triangles, quads) into the parameter space of a nurbs volume.

**Changelog**
- Add SurfaceInNurbsVolumeGeometry.
- Add QuadraturePointSurfaceInNurbsVolumeGeometry.
- Add Container: LOCAL_TANGENT_MATRIX
- Unit Test.

Test relies on #8997.

### CODE REVIEW

File: kratos/geometries/quadrature_point_surface_in_volume_geometry.h [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+    ///@{
+
+    /// Constructor with points and geometry shape function container
+    QuadraturePointSurfaceInVolumeGeometry(
+        const PointsArrayType& ThisPoints,
+        GeometryShapeFunctionContainerType& ThisGeometryShapeFunctionContainer,
+        TangentMatrixType LocalTangents )
+        : BaseType(ThisPoints, ThisGeometryShapeFunctionContainer)
+        , mLocalTangents(LocalTangents)
+    {
```
Comment (tteschemacher) [inline kratos/geometries/quadrature_point_surface_in_volume_geometry.h:73]: you should make some size checks for the local tangents, here.

File: kratos/geometries/quadrature_point_surface_in_volume_geometry.h [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+
+    /// Constructor with points, geometry shape function container and parent
+    QuadraturePointSurfaceInVolumeGeometry(
+        const PointsArrayType& ThisPoints,
+        GeometryShapeFunctionContainerType& ThisGeometryShapeFunctionContainer,
+        TangentMatrixType LocalTangents,
+        GeometryType* pGeometryParent)
+        : BaseType(ThisPoints, ThisGeometryShapeFunctionContainer, pGeometryParent)
+        , mLocalTangents(LocalTangents)
+    {
```
Comment (tteschemacher) [inline kratos/geometries/quadrature_point_surface_in_volume_geometry.h:87]: here as well.

File: kratos/geometries/quadrature_point_surface_in_volume_geometry.h [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+
+        if (rVariable == LOCAL_TANGENT_MATRIX) {
+            mLocalTangents = rInput;
+        }
+    }
+
+    /// Calculate with Matrix
+    void Calculate(
+        const Variable<Matrix >& rVariable,
+        Matrix& rOutput)
```
Comment (tteschemacher) [inline kratos/geometries/quadrature_point_surface_in_volume_geometry.h]: not sure why the CI complains but you must have the `const override` here.

File: kratos/geometries/surface_in_nurbs_volume_geometry.h [hunk 4]
```diff
@@ -0,0 +1,476 @@
...
+    }
+
+    ///@}
+    ///@name Geometrical Informations
+    ///@{
+
+    /// Provides the center of the underlying surface in local space.
+    Point Center() const override
+    {
+        return mpSurface->Center();
```
Comment (tteschemacher) [inline kratos/geometries/surface_in_nurbs_volume_geometry.h]: this is dodgy as f. :P

File: kratos/geometries/surface_in_nurbs_volume_geometry.h [hunk 5]
```diff
@@ -0,0 +1,476 @@
...
+    }
+
+    /// Computes the area of the underlying surface in local space.
+    double Area() const override
+    {
+        return mpSurface->Area();
```
Comment (tteschemacher) [inline kratos/geometries/surface_in_nurbs_volume_geometry.h]: well, yeh nah-.-.-

File: kratos/geometries/surface_in_nurbs_volume_geometry.h [hunk 6]
```diff
@@ -0,0 +1,476 @@
...
+    {
+        std::vector<CoordinatesArrayType> global_space_derivatives(3);
+        this->GlobalSpaceDerivatives( global_space_derivatives, rPoint, 1);
+
+        BoundedMatrix<double,3,2> global_tangents;
+
+        column(global_tangents,0) = global_space_derivatives[1];
+        column(global_tangents,1) = global_space_derivatives[2];
+
+        return MathUtils<double>::GeneralizedDet( global_tangents );
```
Comment (tteschemacher) [inline kratos/geometries/surface_in_nurbs_volume_geometry.h:237]: 🏆

**Comment by tteschemacher** [comment]
generally a really great work. Thank you very much!

except for some error catches and the CI problem it looks good to me.

File: kratos/geometries/surface_in_nurbs_volume_geometry.h (hunk 4 above)
Comment (manuelmessmer) [reply kratos/geometries/surface_in_nurbs_volume_geometry.h]: hehe, well yes could be changed. ;) Well wrong conversion from nurbs_curve_in_surface. BTW the nurbs_curve_in_surface returns the center of the surface... But we should return the center of the curve in global space, right? Here: center of surface in global space?

File: kratos/geometries/quadrature_point_surface_in_volume_geometry.h (hunk 3 above)
Comment (manuelmessmer) [reply kratos/geometries/quadrature_point_surface_in_volume_geometry.h]: Yes, this was just a try. I thought it is related to this.

File: kratos/geometries/surface_in_nurbs_volume_geometry.h (hunk 5 above)
Comment (manuelmessmer) [reply kratos/geometries/surface_in_nurbs_volume_geometry.h]: hehe, it does return an area ;)

File: kratos/geometries/surface_in_nurbs_volume_geometry.h (hunk 4 above)
Comment (tteschemacher) [reply kratos/geometries/surface_in_nurbs_volume_geometry.h]: haha this one is at least for sure within the domain ;)

I reckon transferring this center into global space should be sufficient, here.

File: kratos/utilities/quadrature_points_utility.h [hunk 7]
```diff
@@ -129,6 +130,32 @@ namespace Kratos
...
+                    rPoints,
+                    rShapeFunctionContainer,
+                    LocalTangentMatrix,
+                    pGeometryParent);
+        }
+
+        static GeometryPointerType CreateQuadraturePointSurfaceInVolume(
+            GeometryShapeFunctionContainer<GeometryData::IntegrationMethod>& rShapeFunctionContainer,
+            PointsArrayType rPoints,
+            Matrix LocalTangentMatrix)
```
Comment (manuelmessmer) [inline kratos/utilities/quadrature_points_utility.h:150]: I changed this form BoundedMatrix to Matrix, such that we can catch an dimension error in `quadrature_point_surface_in_nurbs_volume_geometry`. Btw all the points and matrices, we could pass by reference, as they will be copied when the respective constructor is called afterwards anyway. Right now, we are copying twice, right?

File: kratos/geometries/quadrature_point_surface_in_volume_geometry.h (hunk 3 above)
Comment (manuelmessmer) [reply kratos/geometries/quadrature_point_surface_in_volume_geometry.h]: Fixed :punch:

File: kratos/geometries/quadrature_point_surface_in_volume_geometry.h [hunk 8]
```diff
@@ -69,8 +70,11 @@ class QuadraturePointSurfaceInVolumeGeometry
         GeometryShapeFunctionContainerType& ThisGeometryShapeFunctionContainer,
         TangentMatrixType LocalTangents )
         : BaseType(ThisPoints, ThisGeometryShapeFunctionContainer)
-        , mLocalTangents(LocalTangents)
```
Comment (tteschemacher) [inline kratos/geometries/quadrature_point_surface_in_volume_geometry.h]: I think you can still leave this here. Just only add the error message

Comment (manuelmessmer) [reply kratos/geometries/quadrature_point_surface_in_volume_geometry.h]: But then we might get a segmentation fault, before we catch the error.

Comment (tteschemacher) [reply kratos/geometries/quadrature_point_surface_in_volume_geometry.h]: Hmm could be.

File: kratos/utilities/quadrature_points_utility.h (hunk 7 above)
Comment (tteschemacher) [reply kratos/utilities/quadrature_points_utility.h:150]: That is a good question. I think it should be passed by source, here. Also, the name `rPoints`should be either source or just `Points`. But I believe we should test once doing all those functions with passing the points by ref.

Comment (tteschemacher) [reply kratos/utilities/quadrature_points_utility.h:150]: Let's do this in a separate PR though.

**Review by tteschemacher — APPROVED** [review]
From my side, it looks good. Once again many thanks and also thanks for the changes.

**Comment by manuelmessmer** [comment]
Thanks heaps for reviewing!

**Pull request merged**